rocksdb_admin/admin_handler.cpp

```diff
@@ -929,6 +965,167 @@ void AdminHandler::async_tm_addS3SstFilesToDB(
+      kKafkaWatcherName,
+      kafka_consumer_pool,
+      -1); // kafka_init_blocking_consumer_timeout_ms ;
+
+  {
+    std::lock_guard<std::mutex> lock(kafka_watcher_lock_);
+    kafka_watcher_map_[db_name] = kafka_watcher;
+  }
+
+  // With kafka_init_blocking_consumer_timeout_ms set to -1, messages from
+  // replay_timestamp_ms to the current are synchronously consumed. The
+  // calling thread then returns after spawning a new thread to consume
+  // live messages.
+  kafka_watcher->StartWith(
+      replay_timestamp_ms,
+      [](std::shared_ptr<const RdKafka::Message> message,
+          const bool is_replay) {
+    if (message == nullptr) {
+      LOG(ERROR) << "Message nullptr";
+      return;
+    }
+    const int64_t msg_timestamp_secs = GetMessageTimestampSecs(*message);
+
+    // Logs for debugging
+    LOG(ERROR) << "Key " << *message->key() << ", "
+    << "value " << static_cast<const char *>(message->payload()) << ", "
+    << "partition: " << message->partition() << ", "
+    << "offset: " << message->offset() << ", "
+    << "payload len: " << message->len() << ", "
+    << "msg_timestamp: " << ToUTC(msg_timestamp_secs) << " or "
+    << std::to_string(msg_timestamp_secs) << " secs";
+
+    // TODO: add the logic to write the key/value to rocksdb here
+
+    // TODO: periodically write the message timestamp to local_meta_db
+  });
+
+  LOG(INFO) << "Now consuming live messages for " << db_name;
+
+  // Live messages continue to consumed in a separate thread, but release
+  // callback and return so that helix can transition this partition to
+  // bootstrap stage and the admin thread can be freed.
+  callback.release()->result(StartMessageIngestionResponse());
+}
+
+void AdminHandler::async_tm_stopMessageIngestion(
+    std::unique_ptr<apache::thrift::HandlerCallback<std::unique_ptr<
+      StopMessageIngestionResponse>>> callback,
+    std::unique_ptr<StopMessageIngestionRequest> request) {
+
+  admin::AdminException e;
+  e.errorCode = AdminErrorCode::DB_ADMIN_ERROR;
+
+  const auto db_name = request->db_name;
+  LOG(ERROR) << "Called stopMessageIngestion for " << db_name;
+  db_admin_lock_.Lock(db_name);
+  SCOPE_EXIT { db_admin_lock_.Unlock(db_name); };
+  auto db = getDB(db_name, &e);
+  if (db == nullptr) {
+    e.message = db_name + " doesn't exist.";
+    callback.release()->exceptionInThread(std::move(e));
+    LOG(ERROR) << "Database doesn't exist: " << db_name;
+    return;
+  }
+
+  std::shared_ptr<KafkaWatcher> kafka_watcher;
+  {
+    std::lock_guard<std::mutex> lock(kafka_watcher_lock_);
+    auto iter = kafka_watcher_map_.find(db_name);
+    // Verify that there is thread consuming messages for this db.
+    if (iter == kafka_watcher_map_.end()) {
+      e.message = db_name + " is not being consumed";
+      callback.release()->exceptionInThread(std::move(e));
+      LOG(ERROR) << db_name << " is not being currently consumed";
+      return;
+    }
+    kafka_watcher = iter->second;
```

Comment:
nit: we can remove iter from the map here and remove line 1120 to 1123.

Reply:
I would like to remove the entry from the map only after the watcher is successfully stopped. So will keep this as is for now.